{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# read data inputs\n",
    "import numpy as np\n",
    "docword = np.load(\"science2k-doc-word.npy\")\n",
    "worddoc = np.load(\"science2k-word-doc.npy\")\n",
    "with open(\"science2k-titles.txt\") as t:\n",
    "    docs = t.read().splitlines()\n",
    "with open(\"science2k-vocab.txt\") as v:\n",
    "    words = v.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "similarity: 154.15486577\n",
      "dissimilarity: 0.225426913446\n",
      "861.578404097\n",
      "3\n",
      "similarity: 152.456488381\n",
      "dissimilarity: 0.35234881157\n",
      "624.041568009\n",
      "4\n",
      "similarity: 151.847383692\n",
      "dissimilarity: 0.40291756488\n",
      "598.243208093\n",
      "5\n",
      "similarity: 151.267874054\n",
      "dissimilarity: 0.446592182051\n",
      "579.196046265\n",
      "6\n",
      "similarity: 150.82953424\n",
      "dissimilarity: 0.474670435761\n",
      "577.401565436\n",
      "7\n",
      "similarity: 150.679353181\n",
      "dissimilarity: 0.7769653752\n",
      "370.980795927\n",
      "8\n",
      "similarity: 150.14961263\n",
      "dissimilarity: 0.562186624514\n",
      "534.162877887\n",
      "9\n",
      "similarity: 150.0709669\n",
      "dissimilarity: 0.580514953587\n",
      "537.729801154\n",
      "10\n",
      "similarity: 149.820465014\n",
      "dissimilarity: 0.628030525075\n",
      "513.953373626\n",
      "11\n",
      "similarity: 149.423995704\n",
      "dissimilarity: 0.899294091948\n",
      "369.529828422\n",
      "12\n",
      "similarity: 149.086376776\n",
      "dissimilarity: 0.914590237879\n",
      "373.197289447\n",
      "13\n",
      "similarity: 148.925294443\n",
      "dissimilarity: 0.93940898633\n",
      "372.759060004\n",
      "14\n",
      "similarity: 149.034289968\n",
      "dissimilarity: 0.987413717009\n",
      "363.772383223\n",
      "15\n",
      "similarity: 148.921866858\n",
      "dissimilarity: 1.33451931446\n",
      "275.209884336\n",
      "16\n",
      "similarity: 148.377403481\n",
      "dissimilarity: 1.11045540841\n",
      "336.697561304\n",
      "17\n",
      "similarity: 148.526379192\n",
      "dissimilarity: 1.14985020615\n",
      "332.132953057\n",
      "18\n",
      "similarity: 148.176346875\n",
      "dissimilarity: 1.14559025997\n",
      "338.979737752\n",
      "19\n",
      "similarity: 148.295246735\n",
      "dissimilarity: 1.09516421206\n",
      "361.325978815\n",
      "20\n",
      "similarity: 147.908791399\n",
      "dissimilarity: 1.19971956248\n",
      "334.650064549\n",
      "Best k is: 15 for doc-word clustering\n"
     ]
    }
   ],
   "source": [
    "# part a\n",
    "from sklearn.cluster import KMeans\n",
    "import sys\n",
    "import math\n",
    "\n",
    "minScore = sys.float_info.max\n",
    "bestK = 1\n",
    "best_labels = []\n",
    "best_centers = []\n",
    "\n",
    "for i in range(2, 21):\n",
    "    print(i)\n",
    "    l_labels = []\n",
    "    l_centers = []\n",
    "    l_cstds = []\n",
    "    for j in range(10):\n",
    "        kmeans = KMeans(n_clusters=i).fit(docword)\n",
    "        cstd = 0.0\n",
    "        for k in range(len(docword)):\n",
    "            cstd += np.linalg.norm(docword[k] - kmeans.cluster_centers_[kmeans.labels_[k]])\n",
    "        l_labels.append(kmeans.labels_)\n",
    "        l_centers.append(kmeans.cluster_centers_)\n",
    "        l_cstds.append(cstd)\n",
    "    l_labels = np.array(l_labels)\n",
    "    l_centers = np.array(l_centers)\n",
    "    \n",
    "    # choose the best clustering\n",
    "    best_cl = l_cstds.index(min(l_cstds))\n",
    "    labels = l_labels[best_cl]\n",
    "    centers = l_centers[best_cl]\n",
    "    \n",
    "    # initialize score\n",
    "    score = 1\n",
    "    \n",
    "    # firstly consider similarity within same clusters, smaller is better\n",
    "    sim = l_cstds[best_cl] / len(docword)\n",
    "    score *= sim\n",
    "    print(\"similarity: \" + str(sim))\n",
    "    \n",
    "    # secondly consider number of clusters, smaller is better\n",
    "    score *= np.cbrt(i)\n",
    "    \n",
    "    # thirdly consider dissimilarity among centers, larger is better\n",
    "    dissim = np.mean(centers.std(0))\n",
    "    print(\"dissimilarity: \" + str(dissim))\n",
    "    score /= dissim\n",
    "    print(score)\n",
    "    if score < minScore:\n",
    "        minScore = score\n",
    "        bestK = i\n",
    "        best_labels = labels\n",
    "        best_centers = centers\n",
    "\n",
    "print(\"Best k is: \" + str(bestK) + \" for doc-word clustering\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster 1: \n",
      "\"Economic Incentives for Rain Forest Conservation across Scales\"\n",
      "1\n",
      "Cluster 2: \n",
      "\"Polarization of Chemoattractant Receptor Signaling during Neutrophil Chemotaxis\"\n",
      "\"Mitochondrial FtsZ in a Chromophyte Alga\"\n",
      "\"Translocation of Helicobacter pylori CagA into Gastric Epithelial Cells by Type IV Secretion\"\n",
      "\"Requirement of the Inositol Trisphosphate Receptor for Activation of Store-Operated <latex>$Ca^{2+}$</latex> Channels\"\n",
      "\"Distinct Roles of CONSTANS Target Genes in Reproductive Development of Arabidopsis\"\n",
      "\"Proliferation, but Not Growth, Blocked by Conditional Deletion of 40S Ribosomal Protein S6\"\n",
      "\"Perception of Brassinosteroids by the Extracellular Domain of the Receptor Kinase BRI1\"\n",
      "\"Central Role for the Lens in Cave Fish Eye Degeneration\"\n",
      "\"New Insights into an Old Modification\"\n",
      "\"Active Remodeling of Somatic Nuclei in Egg Cytoplasm by the Nucleosomal ATPase ISWI\"\n",
      "187\n",
      "Cluster 3: \n",
      "\"General Acid-Base Catalysis in the Mechanism of a Hepatitis Delta Virus Ribozyme\"\n",
      "\"The Catalytic Pathway of Cytochrome P450cam at Atomic Resolution\"\n",
      "\"Motility Powered by Supramolecular Springs and Ratchets\"\n",
      "\"Structure of the Light-Driven Chloride Pump Halorhodopsin at 1.8 <latex>$\\AA$</latex> Resolution\"\n",
      "\"Structure and Function of a Human <latex>$TAF_{II}250$</latex> Double Bromodomain Module\"\n",
      "\"Structure of the Cytoplasmic b Subunit: T1 Assembly of Voltage-Dependent <latex>$K^+$</latex> Channels\"\n",
      "\"A Structural Model of Transcription Elongation\"\n",
      "\"Twists in Catalysis: Alternating Conformations of Escherichia coli Thioredoxin Reductase\"\n",
      "\"Topologically Linked Protein Rings in the Bacteriophage HK97 Capsid\"\n",
      "\"The Bacterial Flagellar Cap as the Rotary Promoter of Flagellin Self-Assembly\"\n",
      "38\n",
      "Cluster 4: \n",
      "\"Ethanol-Induced Apoptotic Neurodegeneration and Fetal Alcohol Syndrome\"\n",
      "1\n",
      "Cluster 5: \n",
      "\"Corrections and Clarifications: Surveillance and Privacy\"\n",
      "\"Corrections and Clarifications: Thoughts on the Causes of Tree Mortality in Appalachia\"\n",
      "\"Corrections and Clarifications: Nitrogen on the Moon\"\n",
      "\"Corrections and Clarifications: Invasive Plants versus Their New and Old Neighbors: A Mechanism for Exotic Invasion\"\n",
      "\"Corrections and Clarifications: Optically Defined Multifunctional Patterning of Photosensitive Thin-Film Silica Mesophases\"\n",
      "5\n",
      "Cluster 6: \n",
      "\"Dynamics of the Pacific-North American Plate Boundary in the Western United States\"\n",
      "\"Cool Glacial Temperatures and Changes in Moisture Source Recorded in Oman Groundwaters\"\n",
      "\"Widespread Complex Flow in the Interior of the Antarctic Ice Sheet\"\n",
      "\"Isotope Fractionation and Atmospheric Oxygen: Implications for Phanerozoic <latex>$O_2$</latex> Evolution\"\n",
      "\"Satellite Measurements of Sea Surface Temperature through Clouds\"\n",
      "\"High Magma Storage Rates before the 1983 Eruption of Kilauea, Hawaii\"\n",
      "\"Phosphate Depletion in the Western North Atlantic Ocean\"\n",
      "\"Earthquake Potential along the Northern Hayward Fault, California\"\n",
      "\"Recovery and Management Options for Spring/Summer Chinook Salmon in the Columbia River Basin\"\n",
      "\"Reconstruction of the Amazon Basin Effective Moisture Availability over the past 14,000 Years\"\n",
      "112\n",
      "Cluster 7: \n",
      "\"Support Grows for British Exercise to Allocate University Funds\"\n",
      "\"Balancing the Science Budget\"\n",
      "\"Solar Missions Brighten NASA's Hopes for Space Science Research\"\n",
      "\"Surveying the SBIR Program\"\n",
      "\"A Blow to Austria's Scientific Revival\"\n",
      "\"Balancing the Collaboration Equation\"\n",
      "\"Something to Be Done: Treating HIV/AIDS\"\n",
      "\"Help Needed to Rebuild Science in Yugoslavia\"\n",
      "\"Iran's Scientists Cautiously Reach out to the World\"\n",
      "\"I'd like to See America Used as a Global Lab\"\n",
      "46\n",
      "Cluster 8: \n",
      "\"Reaping the Plant Gene Harvest\"\n",
      "\"Genomics: Journey to the Center of Biology\"\n",
      "\"Drug Discovery: A Historical Perspective\"\n",
      "\"The Rise of the Mouse, Biomedicine's Model Mammal\"\n",
      "\"A Mouse Chronology\"\n",
      "\"A Deluge of Patents Creates Legal Hassles for Research\"\n",
      "\"Infectious History\"\n",
      "\"Interfering with Gene Expression\"\n",
      "\"Clones: A Hard Act to Follow\"\n",
      "\"Arabidopsis Comes of Age\"\n",
      "13\n",
      "Cluster 9: \n",
      "\"Dual Signaling Regulated by Calcyon, a D1 Dopamine Receptor Interacting Protein\"\n",
      "\"Kinesin Superfamily Motor Protein KIF17 and mLin-10 in NMDA Receptor-Containing Vesicle Transport\"\n",
      "\"A Myosin I Isoform in the Nucleus\"\n",
      "3\n",
      "Cluster 10: \n",
      "\"The Formation and Early Evolution of the Milky Way Galaxy\"\n",
      "\"A Global View of Martian Surface Compositions from MGS-TES\"\n",
      "\"Galileo at Io: Results from High-Resolution Imaging\"\n",
      "\"Io's Thermal Emission from the Galileo Photopolarimeter-Radiometer\"\n",
      "\"NEAR at Eros: Imaging and Spectral Results\"\n",
      "\"Discovery of Young, Isolated Planetary Mass Objects in the s Orionis Star Cluster\"\n",
      "\"The Fall, Recovery, Orbit, and Composition of the Tagish Lake Meteorite: A New Type of Carbonaceous Chondrite\"\n",
      "\"A Low Temperature Transfer of ALH84001 from Mars to Earth\"\n",
      "\"Solar Wind Record on the Moon: Deciphering Presolar from Planetary Nitrogen\"\n",
      "\"Subduction and Slab Detachment in the Mediterranean-Carpathian Region\"\n",
      "40\n",
      "Cluster 11: \n",
      "\"Can Old Cells Learn New Tricks?\"\n",
      "\"The Business of Stem Cells\"\n",
      "\"Fetal Neuron Grafts Pave the Way for Stem Cell Therapies\"\n",
      "\"Stem Cells in Epithelial Tissues\"\n",
      "\"On the Hunt for a Wolf in Sheep's Clothing\"\n",
      "\"Gene Therapy on Trial\"\n",
      "\"Tracking the Human Fallout from 'Mad Cow Disease'\"\n",
      "\"Synapses Call the Shots\"\n",
      "\"Breaking down Scientific Barriers to the Study of Brain and Mind\"\n",
      "\"A Lifelong Fascination with the Chick Embryo\"\n",
      "25\n",
      "Cluster 12: \n",
      "\"Seismic Images of the Far Side of the Sun\"\n",
      "\"Anomalous Polarization Profiles in Sunspots: Possible Origin of Umbral Flashes\"\n",
      "\"Electron-Induced Inversion of Helical Chirality in Copper Complexes of N,N-Dialkylmethionines\"\n",
      "\"Discovery of a Basaltic Asteroid in the Outer Main Belt\"\n",
      "\"Hierarchical Self-Assembly of F-Actin and Cationic Lipid Complexes: Stacked Three-Layer Tubule Networks\"\n",
      "\"Discovery of a High-Energy Gamma-Ray-Emitting Persistent Microquasar\"\n",
      "\"Electrochemical Micromachining\"\n",
      "\"Imaging the Electron Wave Function in Self-Assembled Quantum Dots\"\n",
      "\"Dimer Preparation That Mimics the Transition State for the Adsorption of <latex>$H_2$</latex> on the Si(100)-2 x 1 Surface\"\n",
      "\"Chemical Etiology of Nucleic Acid Structure: The a-Threofuranosyl-(3'[Rightwards Arrow]2') Oligonucleotide System\"\n",
      "182\n",
      "Cluster 13: \n",
      "\"Memory-A Century of Consolidation\"\n",
      "\"Distal Initiation and Active Propagation of Action Potentials in Interneuron Dendrites\"\n",
      "\"Rapid Extragranular Plasticity in the Absence of Thalamocortical Plasticity in the Developing Primary Visual Cortex\"\n",
      "\"Motion Integration and Postdiction in Visual Awareness\"\n",
      "\"Resetting Central and Peripheral Circadian Oscillators in Transgenic Rats\"\n",
      "\"Psychological and Neural Mechanisms of the Affective Dimension of Pain\"\n",
      "\"Dendritic Computation of Direction Selectivity by Retinal Ganglion Cells\"\n",
      "\"Diversity and Dynamics of Dendritic Signaling\"\n",
      "\"Actin-Based Plasticity in Dendritic Spines\"\n",
      "\"Song Replay during Sleep and Computational Rules for Sensorimotor Vocal Learning\"\n",
      "53\n",
      "Cluster 14: \n",
      "\"Of Chimps and Men\"\n",
      "\"A Tail of Histone Acetylation and DNA Recombination\"\n",
      "\"Bead-Based Fiber-Optic Arrays\"\n",
      "\"Shrewd Investments\"\n",
      "\"Black-Footed Ferret Recovery\"\n",
      "\"Scope of Olefin Polymerization Nickel Catalysts\"\n",
      "\"Corrections and Clarifications: A New Breed of Scientist-Advocate Emerges\"\n",
      "\"Temperature Oscillations in the North Atlantic\"\n",
      "\"Corrections and Clarifications: Molecular Identification of a Eukaryotic, Stretch-Activated Nonselective Cation Channel\"\n",
      "\"Microorganisms Should Be High on DNA Preservation List\"\n",
      "388\n",
      "Cluster 15: \n",
      "\"Mutations in SDHD, A Mitochondrial Complex II Gene, in Hereditary Paraganglioma\"\n",
      "\"Signaling and Circuitry of Multiple MAPK Pathways Revealed by a Matrix of Global Gene Expression Profiles\"\n",
      "\"Gridlock, an HLH Gene Required for Assembly of the Aorta in Zebrafish\"\n",
      "\"Homologs of Small Nucleolar RNAs in Archaea\"\n",
      "\"Evidence for the Evolution of Bdelloid Rotifers without Sexual Reproduction or Genetic Exchange\"\n",
      "\"Genes Expressed in Human Tumor Endothelium\"\n",
      "\"Aureusidin Synthase: A Polyphenol Oxidase Homolog Responsible for Flower Coloration\"\n",
      "\"Intracellular Parasitism by Histoplasma capsulatum: Fungal Virulence and Calcium Dependence\"\n",
      "\"Accumulation of Dietary Cholesterol in Sitosterolemia Caused by Mutations in Adjacent ABC Transporters\"\n",
      "\"The Origins of Genomic Duplications in Arabidopsis\"\n",
      "81\n"
     ]
    }
   ],
   "source": [
    "# get top 10 titles of each cluster\n",
    "for i in range(bestK):\n",
    "    print(\"Cluster \" + str(i+1) + \": \")\n",
    "    dists = []\n",
    "    titles = []\n",
    "    count = 0\n",
    "    for j in range(len(docword)):\n",
    "        if labels[j] == i:\n",
    "            dists.append(np.linalg.norm(docword[j] - centers[i]))\n",
    "            titles.append(docs[j])\n",
    "            count += 1\n",
    "    ranks = [i[0] for i in sorted(enumerate(dists), key=lambda x:x[1])]\n",
    "    for j in range(len(dists)):\n",
    "        if ranks[j] < 10:\n",
    "            print(titles[j])\n",
    "    print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "similarity: 28.3388836446\n",
      "dissimilarity: 0.383219874967\n",
      "93.1704182551\n",
      "3\n",
      "similarity: 28.0272656525\n",
      "dissimilarity: 0.530947875415\n",
      "76.1323544474\n",
      "4\n",
      "similarity: 27.8749498219\n",
      "dissimilarity: 0.619416563878\n",
      "71.4361340191\n",
      "5\n",
      "similarity: 27.7828162479\n",
      "dissimilarity: 0.685139256646\n",
      "69.3405713569\n",
      "6\n",
      "similarity: 27.6996332566\n",
      "dissimilarity: 0.794939208164\n",
      "63.3175134494\n",
      "7\n",
      "similarity: 27.6478819417\n",
      "dissimilarity: 1.27665852541\n",
      "41.427284157\n",
      "8\n",
      "similarity: 27.5921063927\n",
      "dissimilarity: 1.27297832833\n",
      "43.3504731048\n",
      "9\n",
      "similarity: 27.560277788\n",
      "dissimilarity: 0.904111345816\n",
      "63.4077741099\n",
      "10\n",
      "similarity: 27.4757690498\n",
      "dissimilarity: 1.11890559584\n",
      "52.9041504449\n",
      "11\n",
      "similarity: 27.4666689807\n",
      "dissimilarity: 2.03203260215\n",
      "30.0611933601\n",
      "12\n",
      "similarity: 27.3810691698\n",
      "dissimilarity: 1.51233541328\n",
      "41.4504607639\n",
      "13\n",
      "similarity: 27.3667083914\n",
      "dissimilarity: 1.73898407985\n",
      "37.0033811552\n",
      "14\n",
      "similarity: 27.2850284759\n",
      "dissimilarity: 1.7034576764\n",
      "38.6043053608\n",
      "15\n",
      "similarity: 27.2755723466\n",
      "dissimilarity: 1.82469991171\n",
      "36.8648814108\n",
      "16\n",
      "similarity: 27.2369629732\n",
      "dissimilarity: 2.22400938819\n",
      "30.859962343\n",
      "17\n",
      "similarity: 27.2473598144\n",
      "dissimilarity: 2.10928716943\n",
      "33.215313543\n",
      "18\n",
      "similarity: 27.1823639471\n",
      "dissimilarity: 2.02463055904\n",
      "35.1856520541\n",
      "19\n",
      "similarity: 27.164456272\n",
      "dissimilarity: 2.25106235154\n",
      "32.2006539949\n",
      "Best k is: 11 for word-doc clustering\n"
     ]
    }
   ],
   "source": [
    "# part b\n",
    "minScore = sys.float_info.max\n",
    "bestK = 1\n",
    "best_labels = []\n",
    "best_centers = []\n",
    "\n",
    "for i in range(2, 20):\n",
    "    print(i)\n",
    "    l_labels = []\n",
    "    l_centers = []\n",
    "    l_cstds = []\n",
    "    for j in range(10):\n",
    "        kmeans = KMeans(n_clusters=i).fit(worddoc)\n",
    "        cstd = 0.0\n",
    "        for k in range(len(docword)):\n",
    "            cstd += np.linalg.norm(worddoc[k] - kmeans.cluster_centers_[kmeans.labels_[k]])\n",
    "        l_labels.append(kmeans.labels_)\n",
    "        l_centers.append(kmeans.cluster_centers_)\n",
    "        l_cstds.append(cstd)\n",
    "    l_labels = np.array(l_labels)\n",
    "    l_centers = np.array(l_centers)\n",
    "    \n",
    "    # choose the best clustering\n",
    "    best_cl = l_cstds.index(min(l_cstds))\n",
    "    labels = l_labels[best_cl]\n",
    "    centers = l_centers[best_cl]\n",
    "    \n",
    "    # initialize score\n",
    "    score = 1\n",
    "    \n",
    "    # firstly consider similarity within same clusters, smaller is better\n",
    "    sim = l_cstds[best_cl] / len(worddoc)\n",
    "    score *= sim\n",
    "    print(\"similarity: \" + str(sim))\n",
    "    \n",
    "    # secondly consider number of clusters, smaller is better\n",
    "    score *= np.cbrt(i)\n",
    "    \n",
    "    # thirdly consider dissimilarity among centers, larger is better\n",
    "    dissim = np.mean(centers.std(0))\n",
    "    print(\"dissimilarity: \" + str(dissim))\n",
    "    score /= dissim\n",
    "    print(score)\n",
    "    if score < minScore:\n",
    "        minScore = score\n",
    "        bestK = i\n",
    "        best_labels = labels\n",
    "        best_centers = centers\n",
    "\n",
    "print(\"Best k is: \" + str(bestK) + \" for word-doc clustering\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster 1: \n",
      "hour\n",
      "polymerase\n",
      "serum\n",
      "gel\n",
      "deletion\n",
      "endogenous\n",
      "biochem\n",
      "tagged\n",
      "washed\n",
      "electrophoresis\n",
      "110\n",
      "Cluster 2: \n",
      "make\n",
      "people\n",
      "need\n",
      "came\n",
      "dont\n",
      "director\n",
      "look\n",
      "put\n",
      "going\n",
      "thats\n",
      "49\n",
      "Cluster 3: \n",
      "cores\n",
      "scenarios\n",
      "geology\n",
      "geologic\n",
      "anomaly\n",
      "dating\n",
      "warmer\n",
      "tropics\n",
      "basins\n",
      "proxy\n",
      "191\n",
      "Cluster 4: \n",
      "concluded\n",
      "campaign\n",
      "planned\n",
      "equipment\n",
      "prize\n",
      "sort\n",
      "thanks\n",
      "gone\n",
      "intended\n",
      "save\n",
      "591\n",
      "Cluster 5: \n",
      "thank\n",
      "digestion\n",
      "stimulating\n",
      "glycine\n",
      "elicit\n",
      "clonal\n",
      "digested\n",
      "cdnas\n",
      "stably\n",
      "inhibiting\n",
      "722\n",
      "Cluster 6: \n",
      "produced\n",
      "1\n",
      "Cluster 7: \n",
      "likely\n",
      "1\n",
      "Cluster 8: \n",
      "expression\n",
      "binding\n",
      "response\n",
      "activation\n",
      "receptor\n",
      "membrane\n",
      "expressed\n",
      "receptors\n",
      "signaling\n",
      "mediated\n",
      "10\n",
      "Cluster 9: \n",
      "direct\n",
      "cellular\n",
      "functions\n",
      "phosphorylation\n",
      "inhibition\n",
      "induction\n",
      "induce\n",
      "inhibitor\n",
      "regulate\n",
      "activate\n",
      "53\n",
      "Cluster 10: \n",
      "film\n",
      "interface\n",
      "applications\n",
      "oriented\n",
      "schematic\n",
      "scanning\n",
      "deposited\n",
      "films\n",
      "diffraction\n",
      "perpendicular\n",
      "71\n",
      "Cluster 11: \n",
      "fig\n",
      "region\n",
      "larger\n",
      "observations\n",
      "pattern\n",
      "increases\n",
      "relatively\n",
      "significantly\n",
      "estimated\n",
      "decrease\n",
      "116\n"
     ]
    }
   ],
   "source": [
    "# get top 10 words of each cluster\n",
    "for i in range(bestK):\n",
    "    print(\"Cluster \" + str(i+1) + \": \")\n",
    "    dists = []\n",
    "    terms = []\n",
    "    count = 0\n",
    "    for j in range(len(worddoc)):\n",
    "        if labels[j] == i:\n",
    "            dists.append(np.linalg.norm(worddoc[j] - centers[i]))\n",
    "            terms.append(words[j])\n",
    "            count += 1\n",
    "    ranks = [i[0] for i in sorted(enumerate(dists), key=lambda x:x[1])]\n",
    "    for j in range(len(dists)):\n",
    "        if ranks[j] < 10:\n",
    "            print(terms[j])\n",
    "    print(count)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
